core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala

```diff
@@ -0,0 +1,614 @@
+        val memResource = Resource.newBuilder()
+          .setName("mem").setType(Value.Type.SCALAR)
+          .setScalar(Value.Scalar.newBuilder().setValue(driverMem)).build()
+        val commandInfo = buildDriverCommand(submission)
+        val appName = submission.schedulerProperties("spark.app.name")
+        val taskInfo = TaskInfo.newBuilder()
+          .setTaskId(taskId)
+          .setName(s"Driver for $appName")
+          .setSlaveId(offer.offer.getSlaveId)
+          .setCommand(commandInfo)
+          .addResources(cpuResource)
+          .addResources(memResource)
+          .build()
+        val queuedTasks = if (!tasks.contains(offer.offer.getId)) {
+          val buffer = new ArrayBuffer[TaskInfo]
+          tasks(offer.offer.getId) = buffer
+          buffer
+        } else {
+          tasks(offer.offer.getId)
+        }
+        queuedTasks += taskInfo
+        logTrace(s"Using offer ${offer.offer.getId.getValue} to launch driver " +
+          submission.submissionId)
+        val newState = new MesosClusterTaskState(submission, taskId, offer.offer.getSlaveId,
+          None, new Date())
+        launchedDrivers(submission.submissionId) = newState
+        launchedDriversState.persist(submission.submissionId, newState)
+        scheduledCallback(submission.submissionId)
+      }
+    }
+  }
+
+  override def resourceOffers(driver: SchedulerDriver, offers: JList[Offer]): Unit = {
+    /**
+     * Prints the list of Mesos offers for logging purpose.
+     */
+    def printOffers(offers: Iterable[ResourceOffer]): String = {
+      val builder = new StringBuilder()
+      offers.foreach { o =>
+        builder.append(o).append("\n")
+      }
+      builder.toString()
+    }
+    val currentOffers = offers.map { o =>
+      new ResourceOffer(
+        o, getResource(o.getResourcesList, "cpus"), getResource(o.getResourcesList, "mem"))
+    }.toList
+    logTrace(s"Received offers from Mesos: \n${printOffers(currentOffers)}")
+    val tasks = new mutable.HashMap[OfferID, ArrayBuffer[TaskInfo]]()
+    val currentTime = new Date()
+
+    stateLock.synchronized {
+      // We first schedule all the supervised drivers that are ready to retry.
+      // This list will be empty if none of the drivers are marked as supervise.
+      scheduleTasks(pendingRetryDrivers
+        .filter(d => d.retryState.get.nextRetry.before(currentTime)),
+        removeFromPendingRetryDrivers,
+        currentOffers, tasks, currentTime)
+      // Then we walk through the queued drivers and try to schedule them.
+      scheduleTasks(queuedDrivers,
+        removeFromQueuedDrivers,
+        currentOffers, tasks, currentTime)
+    }
+    tasks.foreach { case (offerId, tasks) =>
+      driver.launchTasks(Collections.singleton(offerId), tasks)
+    }
+    offers
+      .filter(o => !tasks.keySet.contains(o.getId))
+      .foreach(o => driver.declineOffer(o.getId))
+  }
+
+  def getState(): MesosClusterSchedulerState = {
```

Comment:
can you call this `getSchedulerState`? There is already a `getStatus` and the names are a little too similar even though they mean completely different things